# Binning EDA

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('CLEANDATA/clean_data.csv', index_col=False)
data.head().T

0  \
TRAFFIC_CONTROL_DEVICE                           TRAFFIC SIGNAL   
DEVICE_CONDITION                           FUNCTIONING PROPERLY   
WEATHER_CONDITION                                         CLEAR   
LIGHTING_CONDITION                                         DUSK   
FIRST_CRASH_TYPE                                          ANGLE   
TRAFFICWAY_TYPE                             FIVE POINT, OR MORE   
ROADWAY_SURFACE_COND                                        DRY   
ROAD_DEFECT                                          NO DEFECTS   
CRASH_TYPE                     INJURY AND / OR TOW DUE TO CRASH   
DAMAGE                                              OVER $1,500   
PRIM_CONTRIBUTORY_CAUSE                     UNABLE TO DETERMINE   
SEC_CONTRIBUTORY_CAUSE                           NOT APPLICABLE   
STREET_NO                                                  5500   
STREET_DIRECTION                                              S   
STREET_NAME                                       WENTWORTH AVE   
BEAT_OF_OCCURRENCE                                        225.0   
NUM_UNITS                                                     2   
MOST_SEVERE_INJURY                        INCAPACITATING INJURY   
INJURIES_TOTAL                                              3.0   
INJURIES_FATAL                                              0.0   
INJURIES_INCAPACITATING                                     1.0   
INJURIES_NON_INCAPACITATING                                 2.0   
INJURIES_REPORTED_NOT_EVIDENT                               0.0   
INJURIES_NO_INDICATION                                      2.0   
INJURIES_UNKNOWN                                            0.0   
CRASH_HOUR                                                   19   
CRASH_DAY_OF_WEEK                                             3   
CRASH_MONTH                                                   9   

                                                        1  \
TRAFFIC_CONTROL_DEVICE                        NO CONTROLS   
DEVICE_CONDITION                              NO CONTROLS   
WEATHER_CONDITION                                   CLEAR   
LIGHTING_CONDITION                 DARKNESS, LIGHTED ROAD   
FIRST_CRASH_TYPE                                 REAR END   
TRAFFICWAY_TYPE                DIVIDED - W/MEDIAN BARRIER   
ROADWAY_SURFACE_COND                                  DRY   
ROAD_DEFECT                                    NO DEFECTS   
CRASH_TYPE                         NO INJURY / DRIVE AWAY   
DAMAGE                                        OVER $1,500   
PRIM_CONTRIBUTORY_CAUSE             FOLLOWING TOO CLOSELY   
SEC_CONTRIBUTORY_CAUSE              FOLLOWING TOO CLOSELY   
STREET_NO                                            7900   
STREET_DIRECTION                                        S   
STREET_NAME                             CHICAGO SKYWAY OB   
BEAT_OF_OCCURRENCE                                  411.0   
NUM_UNITS                                               2   
MOST_SEVERE_INJURY                NO INDICATION OF INJURY   
INJURIES_TOTAL                                        0.0   
INJURIES_FATAL                                        0.0   
INJURIES_INCAPACITATING                               0.0   
INJURIES_NON_INCAPACITATING                           0.0   
INJURIES_REPORTED_NOT_EVIDENT                         0.0   
INJURIES_NO_INDICATION                                2.0   
INJURIES_UNKNOWN                                      0.0   
CRASH_HOUR                                             18   
CRASH_DAY_OF_WEEK                                       6   
CRASH_MONTH                                             9   

                                                                               2  \
TRAFFIC_CONTROL_DEVICE                                            TRAFFIC SIGNAL   
DEVICE_CONDITION                                            FUNCTIONING PROPERLY   
WEATHER_CONDITION                                                          CLEAR   
LIGHTING_CON

In [3]:
data.shape

(882247, 28)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 882247 entries, 0 to 882246
Data columns (total 28 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   TRAFFIC_CONTROL_DEVICE         882247 non-null  object 
 1   DEVICE_CONDITION               882247 non-null  object 
 2   WEATHER_CONDITION              882247 non-null  object 
 3   LIGHTING_CONDITION             882247 non-null  object 
 4   FIRST_CRASH_TYPE               882247 non-null  object 
 5   TRAFFICWAY_TYPE                882247 non-null  object 
 6   ROADWAY_SURFACE_COND           882247 non-null  object 
 7   ROAD_DEFECT                    882247 non-null  object 
 8   CRASH_TYPE                     882247 non-null  object 
 9   DAMAGE                         882247 non-null  object 
 10  PRIM_CONTRIBUTORY_CAUSE        882247 non-null  object 
 11  SEC_CONTRIBUTORY_CAUSE         882247 non-null  object 
 12  STREET_NO                     

## Looking up Binning

In [5]:
# Traffic_Control_Device

data['TRAFFIC_CONTROL_DEVICE'].value_counts()

TRAFFIC_CONTROL_DEVICE
NO CONTROLS                 499451
TRAFFIC SIGNAL              244985
STOP SIGN/FLASHER            87546
UNKNOWN                      36971
OTHER                         5975
YIELD                         1337
LANE USE MARKING              1223
OTHER REG. SIGN               1050
OTHER WARNING SIGN             705
PEDESTRIAN CROSSING SIGN       611
RAILROAD CROSSING GATE         569
FLASHING CONTROL SIGNAL        360
DELINEATORS                    351
SCHOOL ZONE                    341
POLICE/FLAGMAN                 303
OTHER RAILROAD CROSSING        191
RR CROSSING SIGN               188
NO PASSING                      56
BICYCLE CROSSING SIGN           34
Name: count, dtype: int64

In [6]:
(36971 / 882247) * 100

4.190549812014096

In [7]:
# Device_condition

data['DEVICE_CONDITION'].value_counts()

DEVICE_CONDITION
NO CONTROLS                 505406
FUNCTIONING PROPERLY        301784
UNKNOWN                      61412
OTHER                         6674
FUNCTIONING IMPROPERLY        4060
NOT FUNCTIONING               2518
WORN REFLECTIVE MATERIAL       294
MISSING                         99
Name: count, dtype: int64

In [8]:
(61412 / 882247) * 100

6.960862434216268

In [9]:
# weather condition

data['WEATHER_CONDITION'].value_counts()

WEATHER_CONDITION
CLEAR                       694654
RAIN                         76411
UNKNOWN                      49884
SNOW                         28335
CLOUDY/OVERCAST              25573
OTHER                         2715
FREEZING RAIN/DRIZZLE         1738
FOG/SMOKE/HAZE                1345
SLEET/HAIL                    1011
BLOWING SNOW                   441
SEVERE CROSS WIND GATE         133
BLOWING SAND, SOIL, DIRT         7
Name: count, dtype: int64

In [10]:
(49884 / 882247) * 100

5.65419888081229

In [11]:
# lighting condition

data['LIGHTING_CONDITION'].value_counts()

LIGHTING_CONDITION
DAYLIGHT                  568354
DARKNESS, LIGHTED ROAD    191705
DARKNESS                   41246
UNKNOWN                    41206
DUSK                       25093
DAWN                       14643
Name: count, dtype: int64

In [12]:
(41206 / 882247) * 100

4.6705741135985726

In [13]:
# FIRST_CRASH_TYPE

data['FIRST_CRASH_TYPE'].value_counts()

FIRST_CRASH_TYPE
PARKED MOTOR VEHICLE            203997
REAR END                        195887
SIDESWIPE SAME DIRECTION        135546
TURNING                         126908
ANGLE                            96125
FIXED OBJECT                     40573
PEDESTRIAN                       20820
PEDALCYCLIST                     14023
SIDESWIPE OPPOSITE DIRECTION     12297
REAR TO FRONT                     8984
OTHER OBJECT                      8560
HEAD ON                           7513
REAR TO SIDE                      5361
OTHER NONCOLLISION                2580
REAR TO REAR                      1866
ANIMAL                             637
OVERTURNED                         526
TRAIN                               44
Name: count, dtype: int64

In [14]:
# TRAFFICWAY_TYPE

data['TRAFFICWAY_TYPE'].value_counts()

TRAFFICWAY_TYPE
NOT DIVIDED                        380743
DIVIDED - W/MEDIAN (NOT RAISED)    139726
ONE-WAY                            111692
FOUR WAY                            60333
PARKING LOT                         59663
DIVIDED - W/MEDIAN BARRIER          50009
OTHER                               23828
ALLEY                               14477
T-INTERSECTION                      11986
UNKNOWN                             10342
CENTER TURN LANE                     6274
DRIVEWAY                             2844
RAMP                                 2769
UNKNOWN INTERSECTION TYPE            2691
FIVE POINT, OR MORE                  1331
Y-INTERSECTION                       1300
TRAFFIC ROUTE                        1101
NOT REPORTED                          658
ROUNDABOUT                            296
L-INTERSECTION                        184
Name: count, dtype: int64

### combine
* UNKNOWN INTERSECTION TYPE, UNKNOWN and OTHER
* DIVIDED - W/MEDIAN (NOT RAISED), DIVIDED - W/MEDIAN BARRIER
* ROUNDABOUT, FIVE POINT, OR MORE
* T, L, T intersections
* 

### new 
* PARKING LOT
* ALLEY
* NOT REPORTED
* CENTER TURN LANE
* RAMP


In [15]:
(10342 / 882247) * 100

1.1722340795718207

In [16]:
# ROADWAY_SURFACE_COND

data['ROADWAY_SURFACE_COND'].value_counts()

ROADWAY_SURFACE_COND
DRY                653417
WET                115165
UNKNOWN             77420
SNOW OR SLUSH       28058
ICE                  5652
OTHER                2220
SAND, MUD, DIRT       315
Name: count, dtype: int64

### combine
* UNKNOWN, OTHER

In [17]:
(77420 / 882247) * 100

8.775320290122835

In [18]:
# ROAD_DEFECT

data['ROAD_DEFECT'].value_counts()

ROAD_DEFECT
NO DEFECTS           704227
UNKNOWN              161099
RUT, HOLES             6304
OTHER                  4781
WORN SURFACE           3651
SHOULDER DEFECT        1536
DEBRIS ON ROADWAY       649
Name: count, dtype: int64

### Combine
* UNKNOWN, OTHER -> UNKNOWN
* RUT, HOLES, WORN SURFACE -> WORN SURFACE

### New
* NO DEFECT
* UNKNOWN
* WORN SURFACE
* SHOULDER DEFECT
* DEBRIS ON ROADWAY

In [19]:
(161099 / 882247) * 100

18.260079093496493

In [20]:
# CRASH_TYPE

data['CRASH_TYPE'].value_counts()

CRASH_TYPE
NO INJURY / DRIVE AWAY              645376
INJURY AND / OR TOW DUE TO CRASH    236871
Name: count, dtype: int64

In [21]:
# DAMAGE

data['DAMAGE'].value_counts()

DAMAGE
OVER $1,500      553727
$501 - $1,500    229184
$500 OR LESS      99336
Name: count, dtype: int64

In [22]:
# PRIM_CONTRIBUTORY_CAUSE

data['PRIM_CONTRIBUTORY_CAUSE'].value_counts()

PRIM_CONTRIBUTORY_CAUSE
UNABLE TO DETERMINE                                                                 344446
FAILING TO YIELD RIGHT-OF-WAY                                                        97534
FOLLOWING TOO CLOSELY                                                                85376
NOT APPLICABLE                                                                       46487
IMPROPER OVERTAKING/PASSING                                                          43982
FAILING TO REDUCE SPEED TO AVOID CRASH                                               37113
IMPROPER BACKING                                                                     34282
IMPROPER LANE USAGE                                                                  31502
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                                                  29913
IMPROPER TURNING/NO SIGNAL                                                           29570
DISREGARDING TRAFFIC SIGNALS                                      

### Combined
(UNKNOWN)
UNABLE TO DETERMINE                                                                 344446
NOT APPLICABLE                                                                       46487

(WEATHER)
WEATHER                                                                              12613

(RECKLESS_DRIVING)
OPERATING VEHICLE IN ERRATIC, RECKLESS, CARELESS, NEGLIGENT OR AGGRESSIVE MANNER     11113
IMPROPER OVERTAKING/PASSING                                                          43982
IMPROPER BACKING                                                                     34282
IMPROPER LANE USAGE                                                                  31502
IMPROPER TURNING/NO SIGNAL                                                           29570
FOLLOWING TOO CLOSELY                                                                85376
FAILING TO YIELD RIGHT-OF-WAY                                                        97534
DRIVING ON WRONG SIDE/WRONG WAY                                                       4769
EXCEEDING AUTHORIZED SPEED LIMIT                                                      1971

(DISREGARDING_ROADSIGNS)
EXCEEDING SAFE SPEED FOR CONDITIONS                                                   1681
DISREGARDING STOP SIGN                                                                9466
DISREGARDING OTHER TRAFFIC SIGNS                                                      1884
DISREGARDING ROAD MARKINGS                                                            1076
DISREGARDING YIELD SIGN                                                                276
DISREGARDING TRAFFIC SIGNALS                                                         17305
PASSING STOPPED SCHOOL BUS                                                             109
RELATED TO BUS STOP                                                                    456

(DISTRACTED)
DISTRACTION - FROM INSIDE VEHICLE                                                     5994
DISTRACTION - FROM OUTSIDE VEHICLE                                                    3576
CELL PHONE USE OTHER THAN TEXTING                                                     1167
TEXTING                                                                                335
DISTRACTION - OTHER ELECTRONIC DEVICE (NAVIGATION DEVICE, DVD PLAYER, ETC.)            417

(UNDER_INFLUENCE)
UNDER THE INFLUENCE OF ALCOHOL/DRUGS (USE WHEN ARREST IS EFFECTED)                    4100
HAD BEEN DRINKING (USE WHEN ARREST IS NOT MADE)                                        879
PHYSICAL CONDITION OF DRIVER                                                          5203

(ROAD_ENGINEERING)
ROAD ENGINEERING/SURFACE/MARKING DEFECTS                                              2113
ROAD CONSTRUCTION/MAINTENANCE                                                         1857
VISION OBSCURED (SIGNS, TREE LIMBS, BUILDINGS, ETC.)                                  5031
OBSTRUCTED CROSSWALKS                                                                   96
ANIMAL                                                                                 750

(DRIVING_EXPERIENCE)
BICYCLE ADVANCING LEGALLY ON RED LIGHT                                                  86
MOTORCYCLE ADVANCING LEGALLY ON RED LIGHT                                               23
TURNING RIGHT ON RED                                                                   673
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                                                  29913
FAILING TO REDUCE SPEED TO AVOID CRASH                                               37113
EQUIPMENT - VEHICLE CONDITION                                                         5414
EVASIVE ACTION DUE TO ANIMAL, OBJECT, NONMOTORIST                                     1609

In [23]:
# SEC_CONTRIBUTORY_CAUSE	

data['SEC_CONTRIBUTORY_CAUSE'].value_counts()

SEC_CONTRIBUTORY_CAUSE
NOT APPLICABLE                                                                      363455
UNABLE TO DETERMINE                                                                 318000
FAILING TO REDUCE SPEED TO AVOID CRASH                                               32552
FAILING TO YIELD RIGHT-OF-WAY                                                        28264
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                                                  27433
FOLLOWING TOO CLOSELY                                                                23297
IMPROPER OVERTAKING/PASSING                                                          13729
IMPROPER LANE USAGE                                                                  12434
WEATHER                                                                               9705
IMPROPER TURNING/NO SIGNAL                                                            9153
IMPROPER BACKING                                                   

(UNKNOWN)
UNABLE TO DETERMINE                                                                 344446
NOT APPLICABLE                                                                       46487

(WEATHER)
WEATHER                                                                              12613

(RECKLESS_DRIVING)
OPERATING VEHICLE IN ERRATIC, RECKLESS, CARELESS, NEGLIGENT OR AGGRESSIVE MANNER     11113
IMPROPER OVERTAKING/PASSING                                                          43982
IMPROPER BACKING                                                                     34282
IMPROPER LANE USAGE                                                                  31502
IMPROPER TURNING/NO SIGNAL                                                           29570
FOLLOWING TOO CLOSELY                                                                85376
FAILING TO YIELD RIGHT-OF-WAY                                                        97534
DRIVING ON WRONG SIDE/WRONG WAY                                                       4769
EXCEEDING AUTHORIZED SPEED LIMIT                                                      1971

(DISREGARDING_ROADSIGNS)
EXCEEDING SAFE SPEED FOR CONDITIONS                                                   1681
DISREGARDING STOP SIGN                                                                9466
DISREGARDING OTHER TRAFFIC SIGNS                                                      1884
DISREGARDING ROAD MARKINGS                                                            1076
DISREGARDING YIELD SIGN                                                                276
DISREGARDING TRAFFIC SIGNALS                                                         17305
PASSING STOPPED SCHOOL BUS                                                             109
RELATED TO BUS STOP                                                                    456

(DISTRACTED)
DISTRACTION - FROM INSIDE VEHICLE                                                     5994
DISTRACTION - FROM OUTSIDE VEHICLE                                                    3576
CELL PHONE USE OTHER THAN TEXTING                                                     1167
TEXTING                                                                                335
DISTRACTION - OTHER ELECTRONIC DEVICE (NAVIGATION DEVICE, DVD PLAYER, ETC.)            417

(UNDER_INFLUENCE)
UNDER THE INFLUENCE OF ALCOHOL/DRUGS (USE WHEN ARREST IS EFFECTED)                    4100
HAD BEEN DRINKING (USE WHEN ARREST IS NOT MADE)                                        879
PHYSICAL CONDITION OF DRIVER                                                          5203

(ROAD_ENGINEERING)
ROAD ENGINEERING/SURFACE/MARKING DEFECTS                                              2113
ROAD CONSTRUCTION/MAINTENANCE                                                         1857
VISION OBSCURED (SIGNS, TREE LIMBS, BUILDINGS, ETC.)                                  5031
OBSTRUCTED CROSSWALKS                                                                   96
ANIMAL                                                                                 750

(DRIVING_EXPERIENCE)
BICYCLE ADVANCING LEGALLY ON RED LIGHT                                                  86
MOTORCYCLE ADVANCING LEGALLY ON RED LIGHT                                               23
TURNING RIGHT ON RED                                                                   673
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                                                  29913
FAILING TO REDUCE SPEED TO AVOID CRASH                                               37113
EQUIPMENT - VEHICLE CONDITION                                                         5414
EVASIVE ACTION DUE TO ANIMAL, OBJECT, NONMOTORIST                                     1609

In [24]:
# MOST_SEVERE_INJURY

data['MOST_SEVERE_INJURY'].value_counts()

MOST_SEVERE_INJURY
NO INDICATION OF INJURY     758276
NONINCAPACITATING INJURY     69707
REPORTED, NOT EVIDENT        38448
INCAPACITATING INJURY        14849
FATAL                          967
Name: count, dtype: int64

# Binning

In [25]:
data['SEC_CONTRIBUTORY_CAUSE_N'] = data['SEC_CONTRIBUTORY_CAUSE'].apply(
    lambda x: 'UNKNOWN' if x in ['UNABLE TO DETERMINE', 'NOT APPLICABLE'] 
    else 'WEATHER' if x in [
        'WEATHER']
    
    else 'RECKLESS_DRIVING' if x in [
        'OPERATING VEHICLE IN ERRATIC, RECKLESS, CARELESS, NEGLIGENT OR AGGRESSIVE MANNER', 'IMPROPER OVERTAKING/PASSING', 'IMPROPER BACKING', 'IMPROPER LANE USAGE', 'IMPROPER TURNING/NO SIGNAL', 'FOLLOWING TOO CLOSELY', 'FAILING TO YIELD RIGHT-OF-WAY', 'DRIVING ON WRONG SIDE/WRONG WAY', 'EXCEEDING AUTHORIZED SPEED LIMIT']
    
    else 'DISREGARDING_ROADSIGNS' if x in [
        'EXCEEDING SAFE SPEED FOR CONDITIONS', 'DISREGARDING STOP SIGN', 'DISREGARDING OTHER TRAFFIC SIGNS', 'DISREGARDING ROAD MARKINGS', 'DISREGARDING YIELD SIGN', 'DISREGARDING TRAFFIC SIGNALS', 'PASSING STOPPED SCHOOL BUS', 'RELATED TO BUS STOP']
    
    else 'DISTRACTED' if x in [
        'DISTRACTION - FROM INSIDE VEHICLE', 'DISTRACTION - FROM OUTSIDE VEHICLE', 'CELL PHONE USE OTHER THAN TEXTING', 'TEXTING', 'DISTRACTION - OTHER ELECTRONIC DEVICE (NAVIGATION DEVICE, DVD PLAYER, ETC.)']
    
    else 'UNDER_INFLUENCE' if x in [
        'UNDER THE INFLUENCE OF ALCOHOL/DRUGS (USE WHEN ARREST IS EFFECTED)', 'HAD BEEN DRINKING (USE WHEN ARREST IS NOT MADE)', 'PHYSICAL CONDITION OF DRIVER']
    
    else 'ROAD_ENGINEERING' if x in [
        'ROAD ENGINEERING/SURFACE/MARKING DEFECTS', 'ROAD CONSTRUCTION/MAINTENANCE', 'VISION OBSCURED (SIGNS, TREE LIMBS, BUILDINGS, ETC.)', 'OBSTRUCTED CROSSWALKS', 'ANIMAL']
    
    else 'DRIVING_EXPERIENCE' if x in ['BICYCLE ADVANCING LEGALLY ON RED LIGHT', 'MOTORCYCLE ADVANCING LEGALLY ON RED LIGHT', 'TURNING RIGHT ON RED', 'DRIVING SKILLS/KNOWLEDGE/EXPERIENCE', 'FAILING TO REDUCE SPEED TO AVOID CRASH', 'EQUIPMENT - VEHICLE CONDITION', 'EVASIVE ACTION DUE TO ANIMAL, OBJECT, NONMOTORIST']
    
    else 'OTHER'
    )

In [26]:
data['SEC_CONTRIBUTORY_CAUSE_N'].value_counts()

SEC_CONTRIBUTORY_CAUSE_N
UNKNOWN                   681455
RECKLESS_DRIVING          102740
DRIVING_EXPERIENCE         62858
DISREGARDING_ROADSIGNS     10099
WEATHER                     9705
DISTRACTED                  5166
UNDER_INFLUENCE             5124
ROAD_ENGINEERING            5100
Name: count, dtype: int64

In [27]:
data['PRIM_CONTRIBUTORY_CAUSE_N'] = data['PRIM_CONTRIBUTORY_CAUSE'].apply(
    lambda x: 'UNKNOWN' if x in ['UNABLE TO DETERMINE', 'NOT APPLICABLE'] 
    
    else 'WEATHER' if x in [
        'WEATHER']
    
    else 'RECKLESS_DRIVING' if x in [
        'OPERATING VEHICLE IN ERRATIC, RECKLESS, CARELESS, NEGLIGENT OR AGGRESSIVE MANNER', 'IMPROPER OVERTAKING/PASSING', 'IMPROPER BACKING', 'IMPROPER LANE USAGE', 'IMPROPER TURNING/NO SIGNAL', 'FOLLOWING TOO CLOSELY', 'FAILING TO YIELD RIGHT-OF-WAY', 'DRIVING ON WRONG SIDE/WRONG WAY', 'EXCEEDING AUTHORIZED SPEED LIMIT']
    
    else 'DISREGARDING_ROADSIGNS' if x in [
        'EXCEEDING SAFE SPEED FOR CONDITIONS', 'DISREGARDING STOP SIGN', 'DISREGARDING OTHER TRAFFIC SIGNS', 'DISREGARDING ROAD MARKINGS', 'DISREGARDING YIELD SIGN', 'DISREGARDING TRAFFIC SIGNALS', 'PASSING STOPPED SCHOOL BUS', 'RELATED TO BUS STOP']
    
    else 'DISTRACTED' if x in [
        'DISTRACTION - FROM INSIDE VEHICLE', 'DISTRACTION - FROM OUTSIDE VEHICLE', 'CELL PHONE USE OTHER THAN TEXTING', 'TEXTING', 'DISTRACTION - OTHER ELECTRONIC DEVICE (NAVIGATION DEVICE, DVD PLAYER, ETC.)']
    
    else 'UNDER_INFLUENCE' if x in [
        'UNDER THE INFLUENCE OF ALCOHOL/DRUGS (USE WHEN ARREST IS EFFECTED)', 'HAD BEEN DRINKING (USE WHEN ARREST IS NOT MADE)', 'PHYSICAL CONDITION OF DRIVER']
    
    else 'ROAD_ENGINEERING' if x in [
        'ROAD ENGINEERING/SURFACE/MARKING DEFECTS', 'ROAD CONSTRUCTION/MAINTENANCE', 'VISION OBSCURED (SIGNS, TREE LIMBS, BUILDINGS, ETC.)', 'OBSTRUCTED CROSSWALKS', 'ANIMAL']
    
    else 'DRIVING_EXPERIENCE' if x in ['BICYCLE ADVANCING LEGALLY ON RED LIGHT', 'MOTORCYCLE ADVANCING LEGALLY ON RED LIGHT', 'TURNING RIGHT ON RED', 'DRIVING SKILLS/KNOWLEDGE/EXPERIENCE', 'FAILING TO REDUCE SPEED TO AVOID CRASH', 'EQUIPMENT - VEHICLE CONDITION', 'EVASIVE ACTION DUE TO ANIMAL, OBJECT, NONMOTORIST']
    
    else 'OTHER'
)

In [28]:
data['PRIM_CONTRIBUTORY_CAUSE_N'].value_counts()

PRIM_CONTRIBUTORY_CAUSE_N
UNKNOWN                   390933
RECKLESS_DRIVING          340099
DRIVING_EXPERIENCE         74831
DISREGARDING_ROADSIGNS     32253
WEATHER                    12613
DISTRACTED                 11489
UNDER_INFLUENCE            10182
ROAD_ENGINEERING            9847
Name: count, dtype: int64

In [29]:
data['TRAFFICWAY_TYPE_N'] = data['TRAFFICWAY_TYPE'].apply(
    lambda x: 'NOT_DIVIDED' if x in [
        'NOT DIVIDED'
    ]
    
    else 'DIVIDED' if x in [
        'DIVIDED - W/MEDIAN (NOT RAISED)', 'DIVIDED - W/MEDIAN BARRIER'
    ]
    
    else 'ONE_WAY' if x in [
        'ONE-WAY'
    ]
    
    else 'PARKED' if x in [
        'PARKING LOT', 'DRIVEWAY'
    ]
    
    else 'UNKNOWN' if x in [
        'OTHER', 'UNKNOWN', 'TRAFFIC ROUTE'
    ]
    
    else 'ALLEY' if x in [
        'ALLEY'
    ]
    
    else 'INTERSECTION' if x in [
        'T-INTERSECTION', 'Y-INTERSECTION', 'L-INTERSECTION', 'UNKNOWN INTERSECTION TYPE', 'FOUR WAY', 'RAMP'
    ]
    
    else 'CENTER_TURN_LANE' if x in [
        'CENTER TURN LANE'
    ]
    
    else 'ROUNDABOUT' if x in [
        'FIVE POINT, OR MORE', 'ROUNDABOUT'
    ]
    
    else 'NOT_REPORTED' if x in [
        'NOT REPORTED'
    ]
    
    else 'OTHER'
)

In [30]:
data['TRAFFICWAY_TYPE_N'].value_counts()

TRAFFICWAY_TYPE_N
NOT_DIVIDED         380743
DIVIDED             189735
ONE_WAY             111692
INTERSECTION         79263
PARKED               62507
UNKNOWN              35271
ALLEY                14477
CENTER_TURN_LANE      6274
ROUNDABOUT            1627
NOT_REPORTED           658
Name: count, dtype: int64

In [31]:
# road surface

data['ROAD_DEFECT_N'] = data['ROAD_DEFECT'].apply(
    lambda x: 'NO_DEFECTS' if x in [
        'NO DEFECTS'
    ]
    
    else 'UNKNOWN' if x in [
        'UNKNOWN', 'OTHER'
    ]
    
    else 'WORN_SURFACE' if x in [
        'RUT, HOLES', 'WORN SURFACE'
    ]
    
    else 'SHOULDER_DEFECT' if x in [
        'SHOULDER DEFECT'
    ]
    
    else 'DEBRIS' if x in [
        'DEBRIS ON ROADWAY'
    ]
    
    else 'OTHER'
)


In [32]:
data['ROAD_DEFECT_N'].value_counts()

ROAD_DEFECT_N
NO_DEFECTS         704227
UNKNOWN            165880
WORN_SURFACE         9955
SHOULDER_DEFECT      1536
DEBRIS                649
Name: count, dtype: int64

In [33]:
# road surface
data['ROADWAY_SURFACE_COND_N'] = data['ROADWAY_SURFACE_COND'].apply(
    lambda x: 'DRY' if x in [
        'DRY'
    ]
    
    else 'WET' if x in [
        'WET'
    ]
    
    else 'UNKNOWN' if x in [
        'UNKNOWN', 'OTHER'
    ]
    
    else 'SNOW' if x in [
        'SNOW OR SLUSH'
    ]
    
    else 'ICE' if x in [
        'ICE'
    ]
    
    else 'SAND' if x in [
        'SAND, MUD, DIRT'
    ]
    
    else 'OTHER'
)


In [34]:
data['ROADWAY_SURFACE_COND_N'].value_counts()

ROADWAY_SURFACE_COND_N
DRY        653417
WET        115165
UNKNOWN     79640
SNOW        28058
ICE          5652
SAND          315
Name: count, dtype: int64

In [35]:
# first crash
data['FIRST_CRASH_TYPE_N'] = data['FIRST_CRASH_TYPE'].apply(
    lambda x: 'PARKED' if x in [
        'PARKED MOTOR VEHICLE'
    ]
    
    else 'REAR_END' if x in [
        'REAR END', 'REAR TO REAR', 'REAR TO FRONT'
    ]
    
    else 'SIDESWIPE' if x in [
        'SIDESWIPE SAME DIRECTION' , 'SIDESWIPE OPPOSITE DIRECTION'
    ]
    
    else 'ANGLE' if x in [
        'TURNING', 'ANGLE', 'REAR TO SIDE'
    ]
    
    else 'OBJECT' if x in [
        'FIXED OBJECT', 'OTHER OBJECT'
    ]
    
    else 'PEDESTRIAN' if x in [
        'PEDESTRIAN'
    ]
    
    else 'PEDALCYCLIST' if x in [
        'PEDALCYCLIST'
    ]
    
    else 'HEAD_ON' if x in [
        'HEAD ON'
    ]
    
    else 'NON_COLLISION' if x in [
        'OTHER NONCOLLISION', 'OVERTURNED'
    ]
    
    else 'ANIMAL' if x in [
        'ANIMAL'
    ]
    
    else 'TRAIN' if x in [
        'TRAIN'
    ]
    
    else 'OTHER'
)

In [36]:
data['FIRST_CRASH_TYPE_N'].value_counts()

FIRST_CRASH_TYPE_N
ANGLE            228394
REAR_END         206737
PARKED           203997
SIDESWIPE        147843
OBJECT            49133
PEDESTRIAN        20820
PEDALCYCLIST      14023
HEAD_ON            7513
NON_COLLISION      3106
ANIMAL              637
TRAIN                44
Name: count, dtype: int64

In [37]:
data['WEATHER_CONDITION_N'] = data['WEATHER_CONDITION'].apply(
    lambda x: 'CLEAR' if x in [
        'CLEAR'
    ]
    
    else 'RAIN' if x in [
        'RAIN', 'FREEZING RAIN/DRIZZLE'
    ]
    
    else 'UNKNOWN' if x in [
        'UNKNOWN', 'OTHER'
    ]
    
    else 'SNOW' if x in [
        'SNOW'
    ]
    
    else 'CLOUD' if x in [
        'CLOUDY/OVERCAST'
    ]
    
    else 'LOW_VISIBILITY' if x in [
        'FOG/SMOKE/HAZE', 'BLOWING SAND, SOIL, DIRT', 'BLOWING SNOW'
    ]
    
    else 'SLEET' if x in [
        'SLEET/HAIL'
    ]
    
    else 'STRONG_WIND' if x in [
        'SEVERE CROSS WIND GATE'
    ]
    
    else 'OTHER'
)

In [38]:
data['WEATHER_CONDITION_N'].value_counts()

WEATHER_CONDITION_N
CLEAR             694654
RAIN               78149
UNKNOWN            52599
SNOW               28335
CLOUD              25573
LOW_VISIBILITY      1793
SLEET               1011
STRONG_WIND          133
Name: count, dtype: int64

In [39]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 882247 entries, 0 to 882246
Data columns (total 35 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   TRAFFIC_CONTROL_DEVICE         882247 non-null  object 
 1   DEVICE_CONDITION               882247 non-null  object 
 2   WEATHER_CONDITION              882247 non-null  object 
 3   LIGHTING_CONDITION             882247 non-null  object 
 4   FIRST_CRASH_TYPE               882247 non-null  object 
 5   TRAFFICWAY_TYPE                882247 non-null  object 
 6   ROADWAY_SURFACE_COND           882247 non-null  object 
 7   ROAD_DEFECT                    882247 non-null  object 
 8   CRASH_TYPE                     882247 non-null  object 
 9   DAMAGE                         882247 non-null  object 
 10  PRIM_CONTRIBUTORY_CAUSE        882247 non-null  object 
 11  SEC_CONTRIBUTORY_CAUSE         882247 non-null  object 
 12  STREET_NO                     

In [40]:
data.head().T

0  \
TRAFFIC_CONTROL_DEVICE                           TRAFFIC SIGNAL   
DEVICE_CONDITION                           FUNCTIONING PROPERLY   
WEATHER_CONDITION                                         CLEAR   
LIGHTING_CONDITION                                         DUSK   
FIRST_CRASH_TYPE                                          ANGLE   
TRAFFICWAY_TYPE                             FIVE POINT, OR MORE   
ROADWAY_SURFACE_COND                                        DRY   
ROAD_DEFECT                                          NO DEFECTS   
CRASH_TYPE                     INJURY AND / OR TOW DUE TO CRASH   
DAMAGE                                              OVER $1,500   
PRIM_CONTRIBUTORY_CAUSE                     UNABLE TO DETERMINE   
SEC_CONTRIBUTORY_CAUSE                           NOT APPLICABLE   
STREET_NO                                                  5500   
STREET_DIRECTION                                              S   
STREET_NAME                                       WENTWORTH AVE   
BEAT_OF_OCCURRENCE                                        225.0   
NUM_UNITS                                                     2   
MOST_SEVERE_INJURY                        INCAPACITATING INJURY   
INJURIES_TOTAL                                              3.0   
INJURIES_FATAL                                              0.0   
INJURIES_INCAPACITATING                                     1.0   
INJURIES_NON_INCAPACITATING                                 2.0   
INJURIES_REPORTED_NOT_EVIDENT                               0.0   
INJURIES_NO_INDICATION                                      2.0   
INJURIES_UNKNOWN                                            0.0   
CRASH_HOUR                                                   19   
CRASH_DAY_OF_WEEK                                             3   
CRASH_MONTH                                                   9   
SEC_CONTRIBUTORY_CAUSE_N                                UNKNOWN   
PRIM_CONTRIBUTORY_CAUSE_N                               UNKNOWN   
TRAFFICWAY_TYPE_N                                    ROUNDABOUT   
ROAD_DEFECT_N                                        NO_DEFECTS   
ROADWAY_SURFACE_COND_N                                      DRY   
FIRST_CRASH_TYPE_N                                        ANGLE   
WEATHER_CONDITION_N                                       CLEAR   

                                                        1  \
TRAFFIC_CONTROL_DEVICE                        NO CONTROLS   
DEVICE_CONDITION                              NO CONTROLS   
WEATHER_CONDITION                                   CLEAR   
LIGHTING_CONDITION                 DARKNESS, LIGHTED ROAD   
FIRST_CRASH_TYPE                                 REAR END   
TRAFFICWAY_TYPE                DIVIDED - W/MEDIAN BARRIER   
ROADWAY_SURFACE_COND                                  DRY   
ROAD_DEFECT                                    NO DEFECTS   
CRASH_TYPE                         NO INJURY / DRIVE AWAY   
DAMAGE                                        OVER $1,500   
PRIM_CONTRIBUTORY_CAUSE             FOLLOWING TOO CLOSELY   
SEC_CONTRIBUTORY_CAUSE              FOLLOWING TOO CLOSELY   
STREET_NO                                            7900   
STREET_DIRECTION                                        S   
STREET_NAME                             CHICAGO SKYWAY OB   
BEAT_OF_OCCURRENCE                                  411.0   
NUM_UNITS                                               2   
MOST_SEVERE_INJURY                NO INDICATION OF INJURY   
INJURIES_TOTAL                                        0.0   
INJURIES_FATAL                                        0.0   
INJURIES_INCAPACITATING                               0.0   
INJURIES_NON_INCAPACITATING                           0.0   
INJURIES_REPORTED_NOT_EVIDENT                         0.0   
INJURIES_NO_INDICATION                                2.0   
INJURIES_UNKNOWN                                      0.0   
CRASH_HOUR                                             18   
CR

NO CONTROLS                 499451

TRAFFIC SIGNAL              244985
FLASHING CONTROL SIGNAL        360

STOP SIGN/FLASHER            87546
OTHER REG. SIGN               1050
YIELD                         1337
OTHER WARNING SIGN             705
NO PASSING                      56
BICYCLE CROSSING SIGN           34
PEDESTRIAN CROSSING SIGN       611

UNKNOWN                      36971
OTHER                         5975

LANE USE MARKING              1223
DELINEATORS                    351

RAILROAD CROSSING GATE         569
RR CROSSING SIGN               188
OTHER RAILROAD CROSSING        191

SCHOOL ZONE                    341
POLICE/FLAGMAN                 303


In [41]:
data['TRAFFIC_CONTROL_DEVICE'].value_counts()

TRAFFIC_CONTROL_DEVICE
NO CONTROLS                 499451
TRAFFIC SIGNAL              244985
STOP SIGN/FLASHER            87546
UNKNOWN                      36971
OTHER                         5975
YIELD                         1337
LANE USE MARKING              1223
OTHER REG. SIGN               1050
OTHER WARNING SIGN             705
PEDESTRIAN CROSSING SIGN       611
RAILROAD CROSSING GATE         569
FLASHING CONTROL SIGNAL        360
DELINEATORS                    351
SCHOOL ZONE                    341
POLICE/FLAGMAN                 303
OTHER RAILROAD CROSSING        191
RR CROSSING SIGN               188
NO PASSING                      56
BICYCLE CROSSING SIGN           34
Name: count, dtype: int64

In [42]:
data['TRAFFIC_CONTROL_DEVICE_N'] = data['TRAFFIC_CONTROL_DEVICE'].apply(
    lambda x: 'NO_CONTROLS' if x in [
        'NO CONTROLS'
    ]
    
    else 'TRAFFIC_SIGNAL' if x in [
        'TRAFFIC SIGNAL', 'FLASHING CONTROL SIGNAL'
    ]
    
    else 'ROAD_SIGN' if x in [
        'STOP SIGN/FLASHER', 'OTHER REG. SIGN', 'YIELD', 'OTHER WARNING SIGN','NO PASSING', 'BICYCLE CROSSING SIGN', 'PEDESTRIAN CROSSING SIGN'
    ]
    
    else 'UNKNOWN' if x in [
        'UNKNOWN', 'OTHER'
    ]
    
    else 'LANE_MARKING' if x in [
        'LANE USE MARKING', 'DELINEATORS'
    ]
    
    else 'RAILWAY' if x in [
        'RAILROAD CROSSING GATE', 'RR CROSSING SIGN', 'OTHER RAILROAD CROSSING'
    ]
    
    else 'FLAGMAN' if x in [
        'SCHOOL ZONE', 'POLICE/FLAGMAN'
    ]
    
    else 'OTHER'
)

In [43]:
data['TRAFFIC_CONTROL_DEVICE_N'].value_counts()

TRAFFIC_CONTROL_DEVICE_N
NO_CONTROLS       499451
TRAFFIC_SIGNAL    245345
ROAD_SIGN          91339
UNKNOWN            42946
LANE_MARKING        1574
RAILWAY              948
FLAGMAN              644
Name: count, dtype: int64

NO CONTROLS                 505406

FUNCTIONING PROPERLY        301784

UNKNOWN                      61412
OTHER                         6674

FUNCTIONING IMPROPERLY        4060
NOT FUNCTIONING               2518
WORN REFLECTIVE MATERIAL       294

MISSING                         99

In [45]:
data['DEVICE_CONDITION_N'] = data['DEVICE_CONDITION'].apply(
    lambda x: 'NO_CONTROLS' if x in [
        'NO CONTROLS'
    ]
    
    else 'FUNCTIONING' if x in [
        'FUNCTIONING PROPERLY'
    ]
    
    else 'UNKNOWN' if x in [
        'UNKNOWN', 'OTHER'
    ]
    
    else 'NOT_FUNCTIONING' if x in [
        'FUNCTIONING IMPROPERLY', 'NOT FUNCTIONING', 'WORN REFLECTIVE MATERIAL'
    ]
    
    else 'MISSING' if x in [
        'MISSING'
    ]
    
    else 'OTHER'
)

In [46]:
data['DEVICE_CONDITION_N'].value_counts()

DEVICE_CONDITION_N
NO_CONTROLS        505406
FUNCTIONING        301784
UNKNOWN             68086
NOT_FUNCTIONING      6872
MISSING                99
Name: count, dtype: int64

In [47]:
data.to_csv('CLEANDATA/clean_data_N.csv', index=False)